src/ir/auth_logic/lowering_ast_datalog.cc

```diff
@@ -262,11 +273,180 @@ std::vector<datalog::Rule> LoweringToDatalogPass::QueriesToDLIR(
+            arguments));
+  }
+  return says_extended_relational_declarations;
+}
+
+std::vector<datalog::RelationDeclaration>
+LoweringToDatalogPass::RelationDeclarationToDLIR(
+    const std::vector<datalog::RelationDeclaration>& relation_declarations) {
+  // Note that the additional relation declarations that are generated for
+  // (possibly nested) cansay expressions are based on an environment
+  // that tracks the maximum depth with which they appear. This environment is
+  // populated by side-effect while translating the program body. As a result,
+  // it is assumed that this function is called after the body is translated.
+
+  std::vector<datalog::RelationDeclaration> transformed_declarations =
+      LoweringToDatalogPass::TransformDeclarations(relation_declarations);
+
+  // Produce a mapping from predicate names to predicate typings
+  // (where a predicate typing is the same as a predicate relation declaration)
+  absl::flat_hash_map<std::string_view, datalog::RelationDeclaration>
+      type_environment;
+  for (const datalog::RelationDeclaration& declaration :
+       transformed_declarations) {
+    type_environment.insert({declaration.relation_name(), declaration});
+  }
+
+  std::vector<datalog::RelationDeclaration> can_say_declarations =
+      LoweringToDatalogPass::CanSayDeclarations(type_environment);
+  absl::c_copy(can_say_declarations,
+               std::back_inserter(transformed_declarations));
+
+  // The transformed relation declarations are the same as the ones from the
+  // surface program plus another one for "canActAs"
+  transformed_declarations.push_back(datalog::RelationDeclaration(
+      "canActAs", false,
+      {datalog::Argument(
+           "p1", datalog::ArgumentType(datalog::ArgumentType::Kind::kPrincipal,
+                                       "Principal")),
+       datalog::Argument(
+           "p2", datalog::ArgumentType(datalog::ArgumentType::Kind::kPrincipal,
+                                       "Principal"))}));
+  // Another declaration "isNumber" added aa a work around for
+  // universe_translations. To be removed after adding parser for
+  // universe_translations
+  transformed_declarations.push_back(datalog::RelationDeclaration(
+      "isNumber", false,
+      {datalog::Argument(
+          "x", datalog::ArgumentType(datalog::ArgumentType::Kind::kNumber,
+                                     "Number"))}));
+  // The translated declarations are all extended with "says_" and a speaker
+  // argument
+  std::vector<datalog::RelationDeclaration>
+      says_extended_relational_declarations =
+          LoweringToDatalogPass::SaysExtendRelationDeclarations(
+              transformed_declarations);
+  return says_extended_relational_declarations;
+}
+
+std::vector<datalog::RelationDeclaration>
+LoweringToDatalogPass::QueryRelationDeclarationToDLIR(
+    const std::vector<Query>& queries) {
+  std::vector<datalog::RelationDeclaration> query_declarations =
```

Comment:
Add a comment like this?

```C++
// For every query add the following declaration:
//   .decl query_name(dummy_param: DummyType)
```

Is my understanding of what is generated correct?

Reply:
Yes that is correct.